bgpd/bgp_bfd_vty.c

```diff
@@ -0,0 +1,674 @@
+	jsonstr = json_object_to_json_string_ext(msg, BFDD_JSON_CONV_OPTIONS);
+
+	rv = bfd_control_call(&bac, BMT_NOTIFY_ADD, jsonstr, strlen(jsonstr));
+	json_object_put(msg);
+
+	if (rv == -1) {
+		zlog_debug("%s:%d: monitor add failure", __func__, __LINE__);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int bfdd_monitor_peer(struct peer *peer, const char *label,
+			     bool multihop)
+{
+	struct bgp_peer_notification *bpn;
+	struct peer_group *group;
+	struct listnode *node;
+	int error = 0;
+
+	/* Save notification configuration. */
+	bpn = bpn_find(peer);
+	if (bpn == NULL) {
+		bpn = bpn_new(peer);
+		if (bpn == NULL)
+			return -1;
+	}
+
+	if (CHECK_FLAG(peer->sflags, PEER_STATUS_GROUP)) {
+		group = peer->group;
+		for (ALL_LIST_ELEMENTS_RO(group->peer, node, peer))
+			error |= _bfdd_monitor_peer(peer, label, multihop);
+	} else
+		error = _bfdd_monitor_peer(peer, label, multihop);
+
+	return error;
+}
+
+
+/*
+ * Commands
+ */
+static int use_multihop(const int argc, struct cmd_token *argv[])
+{
+	if (argc == 0)
+		return 0;
+
+	if (argv[argc - 1]->arg && strmatch(argv[argc - 1]->text, "multihop"))
+		return 1;
+
+	return 0;
+}
+
+DEFUN(bfd_monitor_peer, bfd_monitor_peer_cmd,
+      "neighbor <A.B.C.D|X:X::X:X|WORD> bfdd [multihop]",
+      NEIGHBOR_STR NEIGHBOR_ADDR_STR2
+      "Enable BFD support\n"
+      "Use multihop\n")
+{
+	struct peer *peer;
+
+	peer = peer_and_group_lookup_vty(vty, argv[1]->arg);
+	if (peer == NULL)
+		return CMD_WARNING_CONFIG_FAILED;
+
+	if (bfdd_monitor_peer(peer, NULL, use_multihop(argc, argv)) != 0) {
+		vty_out(vty, "%% Failed to configure BFD peer notification.\n");
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+
+	return CMD_SUCCESS;
+}
+
+DEFUN(bfd_monitor_peer_label, bfd_monitor_peer_label_cmd,
+      "neighbor <A.B.C.D|X:X::X:X|WORD> bfdd label WORD",
+      NEIGHBOR_STR NEIGHBOR_ADDR_STR2
+      "Enable BFD support\n"
+      "Use BFD peer with label\n"
+      "Peer label\n")
+{
+	struct peer *peer;
+
+	peer = peer_and_group_lookup_vty(vty, argv[1]->arg);
+	if (peer == NULL)
+		return CMD_WARNING_CONFIG_FAILED;
+
+	if (bfdd_monitor_peer(peer, argv[4]->arg, false) != 0) {
+		vty_out(vty, "%% Failed to configure BFD peer notification.\n");
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+
+	return CMD_SUCCESS;
+}
+
+DEFUN(bfd_unmonitor_peer, bfd_unmonitor_peer_cmd,
+      "no neighbor <A.B.C.D|X:X::X:X|WORD> bfdd",
+      NO_STR NEIGHBOR_STR NEIGHBOR_ADDR_STR2
+      "Configure Bidirectional Forwarding Detection\n")
+{
+	struct peer *peer;
+
+	peer = peer_and_group_lookup_vty(vty, argv[2]->arg);
+	if (peer == NULL)
+		return CMD_WARNING_CONFIG_FAILED;
+
+	if (bfdd_unmonitor_peer(peer) != 0) {
+		vty_out(vty, "%% Failed to configure BFD peer notification.\n");
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+
+	return CMD_SUCCESS;
+}
+
+
+/*
+ * BFD daemon adapter code.
+ */
+static struct bgp_peer_notification *bpn_new(struct peer *p)
+{
+	struct bgp_peer_notification *bpn;
+
+	bpn = calloc(1, sizeof(*bpn));
```

Comment:
Here and in several other places I think it would be better to use `XCALLOC/XMALLOC` instead.